rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,1053 @@
+
+def discriminator_step(inputs):
+  image, noise = inputs
+  gan_output = gan.connect(image, noise)
+  disc_loss, disc_vars = gan_output.discriminator_loss_and_vars()
+  disc_train_op = disc_optimizer.minimize(disc_loss, var_list=disc_vars)
+
+  with tf.control_dependencies([disc_train_op]):
+    return tf.identity(disc_loss)
+
+def generator_step(inputs):
+  image, noise = inputs
+  gan_output = gan.connect(image, noise)
+  gen_loss, gen_vars = gan_output.generator_loss_and_vars()
+  gen_train_op = gen_optimizer.minimize(gen_loss, var_list=gen_vars)
+
+  with tf.control_dependencies([gen_train_op]):
+    return tf.identity(gen_loss)
+
+input_iterator = strategy.make_input_iterator(input_fn)
+per_replica_disc_losses = strategy.run(discriminator_step, input_iterator)
+per_replica_gen_losses = strategy.run(generator_step, input_iterator)
+mean_disc_loss = strategy.reduce(per_replica_disc_losses)
+mean_gen_loss = strategy.reduce(per_replica_gen_losses)
+
+with tf.Session() as session:
+  session.run(strategy.initialize())
+  session.run(input_iterator.initialize())
+  for _ in range(num_train_steps):
+    for _ in range(num_disc_steps):
+      disc_loss = session.run(mean_disc_loss)
+    for _ in range(num_gen_steps):
+      gen_loss = session.run(mean_gen_loss)
+  session.run(strategy.finalize())
+```
+
+### Reinforcement Learning
+
+This is an example of
+[IMPALA](https://deepmind.com/blog/impala-scalable-distributed-deeprl-dmlab-30/)-like
+Reinforcement Learning system, converted to eager style.
+
+```python
+tf.enable_eager_execution()
+
+with strategy.scope():
+  agent = Agent(num_actions, hidden_size, entropy_cost, baseline_cost)
+  optimizer = tf.train.RMSPropOptimizer(learning_rate)
+
+# Queues of trajectories from actors.
+queues = []
+def learner_input(ctx):
+  del ctx  # Unused.
+  queue = tf.FIFOQueue(
+      capacity=1, dtypes=trajectory_dtypes, shapes=trajectory_shapes)
+  queues.append(queue)
+
+  def dequeue_batch():
+    batch = [Transition(*queue.dequeue()) for _ in range(batch_size_per_replica)]
+    # Stack the `Tensor` nests along axis 1.
+    return tf.nest.map_structure(lambda *xs: tf.stack(xs, axis=1), *batch)
+  return dequeue_batch
+
+def learner_step(trajectories):
+  loss = tf.reduce_sum(agent.compute_loss(trajectories))
+  agent_vars = agent.get_all_variables()
+  optimizer.minimize(loss, var_list=agent_vars)
+  return loss, agent_vars
+
+# Create learner inputs.
+learner_inputs = strategy.make_input_iterator(learner_input)
+
+def run_actor(actor_id):
+  queue = queues[actor_id % len(queues)]
+  for _ in range(num_trajectories_per_actor):
+    observation = get_observation_from_environment()
+    action_taken, logits = agent(tf.expand_dims(observation, axis=0))
+    trajectory = Transition(observation, action_taken, logits)
+    queue.enqueue(tf.nest.flatten(trajectory))
+
+# Start the actors.
+for actor_id in range(num_actors):
+  threading.Thread(target=run_actor, args=(actor_id,)).start()
```

Comment:
It would be nice to have a reinforcement learning story where the environments run in a `Process` instead of a `Thread` to support heavier computation.

The [agent model call](https://github.com/tensorflow/community/pull/25/files#diff-cb12a9548a2d42da420cb33feb2f7742R932) and [FIFOQueue.enqueue](https://github.com/tensorflow/community/pull/25/files#diff-cb12a9548a2d42da420cb33feb2f7742R934) would be problematic in a `Process`.

Reply:
Thanks for the reply @petebu. I apologize if I am misunderstanding -- but can the Distribution Strategy API support mirrored variables on the same machine across processes, so that the end user does not need to manage the cross-process weight transfer manually?

Reply:
Similar question to jperl@: When you say "To transfer the weights, you can serialize them out of graph and send them to the other process." does DS support weight transfer between two processes (running on one or multiple machines)? If not, can you outline how DS would be used in combination with other communication libraries to implement distributed RL with distributed data collection and one or more learners (with replay buffers/FIFO queue for experience trajectories)?

Reply:
Is there any update on this? Currently trying to do this (have data gather loop in its own process) and have been using tf.train.replica_device_setter() building the graph in multiple processes and using the parameter server to share weights between agent collecting data and updater that updates the agent's policy model.